File: dotnet/Shop/Pricing/Pricing.cs
namespace Shop.Pricing;

public enum CustomerTier { Standard, Gold }

public sealed record DiscountPolicy(Money Threshold, int Percent);

public sealed record Priced(Money Subtotal, Money Discount, Money Shipping)
{
    public Money Total => new(Subtotal.Cents - Discount.Cents + Shipping.Cents);
}

public static class Pricing
{
    public static readonly Money FreeShippingThreshold = new(5_000);
    public static readonly Money StandardShipping = new(395);

    /// <summary>Free at or above the threshold. The boundary is the point.</summary>
    public static Money ShippingFor(Money subtotal, CustomerTier tier) =>
        tier == CustomerTier.Gold || subtotal.Cents >= FreeShippingThreshold.Cents
            ? Money.Zero
            : StandardShipping;

    public static Priced Apply(Money subtotal, CustomerTier tier, DiscountPolicy policy)
    {
        var discount = subtotal.Cents >= policy.Threshold.Cents
            ? new Money((int)Math.Round(subtotal.Cents * (policy.Percent / 100m), MidpointRounding.AwayFromZero))
            : Money.Zero;

        var discounted = new Money(subtotal.Cents - discount.Cents);
        return new Priced(subtotal, discount, ShippingFor(discounted, tier));
    }
}

File: java/src/main/java/shop/pricing/Pricing.java
package shop.pricing;

import java.util.List;

/** https://endtoendtester.com/languages/java */
public final class Pricing {
    public static final Money FREE_SHIPPING_THRESHOLD = new Money(5_000);
    public static final Money STANDARD_SHIPPING = new Money(395);

    public enum Tier { STANDARD, GOLD }

    public record Line(String sku, int unitCents, int quantity) {
        public Money total() {
            return new Money(unitCents).times(quantity);
        }
    }

    private Pricing() {
    }

    public static Money subtotal(List<Line> lines) {
        return lines.stream().map(Line::total).reduce(Money.ZERO, Money::plus);
    }

    /** Free at or above the threshold. The boundary is the point. */
    public static Money shippingFor(Money subtotal, Tier tier) {
        if (tier == Tier.GOLD || subtotal.cents() >= FREE_SHIPPING_THRESHOLD.cents()) {
            return Money.ZERO;
        }
        return STANDARD_SHIPPING;
    }

    public static Money total(List<Line> lines, Tier tier) {
        Money subtotal = subtotal(lines);
        return subtotal.plus(shippingFor(subtotal, tier));
    }
}
